### PR TITLE
Add uninstall test

### PR DESCRIPTION
**Pull request by blaisedias**
For CI we want to use a single cluster to install and run
builds for different commits.
So the sequence would be
1. Push the images to the test registry
2. Run the install test
3. Run all other tests
4. Run the uninstall test
5. and maybe restart the test registry

This change adds the uninstall test

### CODE REVIEW

File: mayastor-test/e2e/uninstall/uninstall_test.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+// Encapsulate the logic to find where the deploy yamls are
+func getDeployYamlDir() string {
+	_, filename, _, _ := runtime.Caller(0)
+	return path.Clean(filename + "/../../../../deploy")
+}
+
+// Helper for passing yaml from the deploy directory to kubectl
+func deleteDeployYaml(filename string) {
```
Comment (chriswldenyer) [inline mayastor-test/e2e/uninstall/uninstall_test.go:80]: I've just written a function exactly like this. Maybe we need we should gather commonly-used functions together.

**Review by chriswldenyer — APPROVED** [review]

File: mayastor-test/e2e/uninstall/uninstall_test.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+var cfg *rest.Config
+var k8sClient client.Client
+var k8sManager ctrl.Manager
+var testEnv *envtest.Environment
+
+/// Examine the nodes in the k8s cluster and return
+/// the IP address of the master node (if one exists),
+/// The assumption is that the test-registry is accessible via the IP addr of the master,
+/// or any node in the cluster if the master noe does not exist
```
Comment (paulyoong) [inline mayastor-test/e2e/uninstall/uninstall_test.go:36]: typo - "noe" should be "node"

File: mayastor-test/e2e/uninstall/uninstall_test.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	return nodeIPs[0], nil
+}
+
+// Encapsulate the logic to find where the deploy yamls are
+func getDeployYamlDir() string {
+	_, filename, _, _ := runtime.Caller(0)
+	return path.Clean(filename + "/../../../../deploy")
+}
+
+// Helper for passing yaml from the deploy directory to kubectl
```
Comment (paulyoong) [inline mayastor-test/e2e/uninstall/uninstall_test.go:79]: Is this comment meant to be here? It doesn't seem to relate to the delete function.

File: mayastor-test/e2e/uninstall/uninstall_test.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	deleteTemplatedYaml("moac-deployment.yaml.template", "moac", registryAddress)
+	deleteTemplatedYaml("csi-daemonset.yaml.template", "mayastor-csi", registryAddress)
+	deleteDeployYaml("nats-deployment.yaml")
+	deleteDeployYaml("mayastorpoolcrd.yaml")
+	deleteDeployYaml("moac-rbac.yaml")
+	deleteDeployYaml("storage-class.yaml")
+	deleteDeployYaml("namespace.yaml")
+
+	// Given the yamls and the environment described in the test readme,
+	// we expect mayastor to be running on exactly 3 nodes.
```
Comment (paulyoong) [inline mayastor-test/e2e/uninstall/uninstall_test.go:132]: Copy-paste error? We are tearing down Mayastor, so we don't expect it to be running on 3 nodes.

File: mayastor-test/e2e/uninstall/uninstall_test.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+// We expect this to fail a few times before it succeeds,
+// so no throwing errors from here.
+func mayastorReadyPodCount() int {
+	var mayastorDaemonSet appsv1.DaemonSet
+	if k8sClient.Get(context.TODO(), types.NamespacedName{Name: "mayastor", Namespace: "mayastor"}, &mayastorDaemonSet) != nil {
+		return -1
+	}
+	return int(mayastorDaemonSet.Status.CurrentNumberScheduled)
+}
```
Comment (paulyoong) [inline mayastor-test/e2e/uninstall/uninstall_test.go:114]: As Chris already mentioned, I think functions like this should eventually go in a common file, as we are likely to need them in other tests.

**Review by paulyoong — APPROVED** [review]

File: mayastor-test/e2e/uninstall/uninstall_test.go (hunk 4 above)
Comment (blaisedias) [reply mayastor-test/e2e/uninstall/uninstall_test.go]: Done.

File: mayastor-test/e2e/uninstall/uninstall_test.go (hunk 2 above)
Comment (blaisedias) [reply mayastor-test/e2e/uninstall/uninstall_test.go]: Done

File: mayastor-test/e2e/uninstall/uninstall_test.go (hunk 3 above)
Comment (blaisedias) [reply mayastor-test/e2e/uninstall/uninstall_test.go:79]: Yes. It matches the comment for a similar function is the install test.

File: mayastor-test/e2e/uninstall/uninstall_test.go (hunk 5 above)
Comment (blaisedias) [reply mayastor-test/e2e/uninstall/uninstall_test.go:114]: Sure, I will do that as another PR, there is a bunch of stuff that needs to be moved from the install test as well.

File: mayastor-test/e2e/uninstall/uninstall_test.go (hunk 1 above)
Comment (blaisedias) [reply mayastor-test/e2e/uninstall/uninstall_test.go:80]: Sure I will do that as another PR, as there is stuff that needs to be shared from the install test as well.

**Comment by blaisedias** [comment]
bors merge

**Comment by blaisedias** [comment]
bors merge

**Comment by blaisedias** [comment]
bors merge

**Pull request merged**